chore: refactor ssh listener server

kian99 opened the pull request:
## Description
I found that the ssh server in jimm could be improved in some ways:
- Use an in-memory buffered connection instead of listening on a system port. Similar to our Juju tests.
- The internal ssh package was importing internal/jimm when it shouldn't.
- Refactored the sshManager interface so that dialing the Juju controller's SSH server can be mocked.
- Moved the limitListener to a separate package.
- Removed unused functions thanks to the changes above.

## Engineering checklist
<!-- *Check only items that apply* -->

- [ ] Documentation updated
- [x] Covered by unit tests
- [ ] Covered by integration tests

kian99 committed:
chore: refactor ssh listener server

I found that the ssh server in jimm could be improved in some ways:
- Use an in-memory buffered connection instead of listening on a system port. Similar to our Juju tests.
- The internal ssh package was importing internal/jimm when it shouldn't.
- Refactored the sshManager interface so that dialing the Juju controller's SSH server can be mocked.
- Moved the limitListener to a separate package.
- Removed unused functions thanks to the changes above.

SimoneDutto commented on internal/ssh/limitlistener/listener.go:
```diff
@@ -38,10 +38,10 @@ import (
 // connections from the provided Listener, and it timeouts when the max
 // has been reached and no seats has been freed for the timeout period.
-func limitListenerWithTimeout(l net.Listener, n int, timeout time.Duration) net.Listener {
+func ListenerWithTimeout(l net.Listener, n int, timeout time.Duration) net.Listener {
```
```suggestion
func ListenerWithLimitAndTimeout(l net.Listener, n int, timeout time.Duration) net.Listener {
```
I think it's better to express that there is both limit and timeout.

kian99 replied:
I think that's covered by the package name. Wdyt?

SimoneDutto commented on internal/ssh/ssh.go:
```diff
@@ -144,52 +142,36 @@ func directTCPIPHandler(sshManager SSHManager) func(srv *ssh.Server, conn *gossh
 		}
-		// gossh.Request are requests sent outside of the normal stream of data (ex. pty-req for an interactive session).
-		// Since we only need the raw data to redirect, we can discard them.
 		go gossh.DiscardRequests(reqs)
```
why did you remove this comment? is it inaccurate or something?

kian99 replied:
No that was a bit of an accident, I thought I would be removing the `go gossh.DiscardRequests(reqs)` line. I'll re-add it.

SimoneDutto commented on internal/ssh/ssh.go:
```diff
@@ -144,52 +142,36 @@ func directTCPIPHandler(sshManager SSHManager) func(srv *ssh.Server, conn *gossh
-			}
+			defer clientConn.Close()
+			defer controllerConn.Close()
+			_, _ = io.Copy(clientConn, controllerConn)
```
why ignoring the error here?

kian99 replied:
Ah this is mb, I thought io.Copy returns an EOF error but I see the docstring says,
```
// A successful Copy returns err == nil, not err == EOF.
// Because Copy is defined to read from src until EOF, it does
// not treat an EOF from Read as an error to be reported.
```
So I'll revert this.

kian99 replied:
Although I think the way we handle the error is wrong, we can't reject the channel after accepting it.

SimoneDutto commented on internal/ssh/ssh.go:
```diff
@@ -144,52 +142,36 @@ func directTCPIPHandler(sshManager SSHManager) func(srv *ssh.Server, conn *gossh
-			}
+			defer clientConn.Close()
+			defer controllerConn.Close()
+			_, _ = io.Copy(controllerConn, clientConn)
```
why ignoring the error here?

kian99 replied:
Same as above.

SimoneDutto commented on internal/ssh/ssh_test.go:
```diff
@@ -245,51 +258,27 @@ func (s *sshSuite) TestSSHJumpDialFail(c *qt.C) {
 	c.Assert(err, qt.ErrorMatches, ".*connect failed.*")
 }
 
-func (s *sshSuite) TestMaxConcurrentConnections(c *qt.C) {
```
i guess this test is to be removed with the in-memory connection right?

SimoneDutto replied:
i don't feel like using an untested listener in our ssh server. Shouldn't we try to build a test for it?

kian99 replied:
I realised that we probably don't want to test the listener in the SSH package. We also didn't copy the listeners tests when we copied the listener, so I'm working on a follow-up to copy the listener's existing tests + a new one for the timeout.

kian99 replied:
Added the tests in #1550

SimoneDutto commented on internal/jimm/jimm.go:
```diff
@@ -195,9 +196,8 @@ type SSHManager interface {
-	// a valid JWT To connect to the controller.
-	ControllerInfoFromModelUUID(ctx context.Context, modelUUID string, user *openfga.User) (ssh.ControllerInfo, error)
+	// DialControllerSSHServer dials the controller hosting the specific model UUID.
+	DialControllerSSHServer(ctx context.Context, modelUUID string, user *openfga.User) (*gossh.Client, error)
```
i preferred to dial the in `internal/ssh` instead of `internal/jimm/ssh`, but it's just a preference. I am curious to listen to other opinions

kian99 replied:
This let's us mock the dial for testing so that's one reason.

SimoneDutto commented on internal/ssh/ssh_test.go:
```diff
@@ -217,17 +226,21 @@ func (s *sshSuite) TestSSHJumpPermissionFail(c *qt.C) {
 	for _, test := range tests {
 		c.Run(test.name, func(c *qt.C) {
-			client, err := gossh.Dial("tcp", fmt.Sprintf(":%d", s.jumpServerPort), &gossh.ClientConfig{
+			conn, err := s.jumpServerListener.Dial()
+			c.Assert(err, qt.IsNil)
+
+			sshConn, newChan, reqs, err := gossh.NewClientConn(conn, "", &gossh.ClientConfig{
```
can you create a private method in this file called `dialInMemory` or something, so you can explain why you are using these three functions to dial + you can reuse it in all tests, avoiding this boilerplate?

kian99 replied:
Yes good idea

SimoneDutto commented on internal/testutils/jimmtest/utils.go:
```diff
@@ -1,20 +0,0 @@
-)
-
-// GetFreePort asks the kernel for a free open port that is ready to use.
-func GetFreePort() (int, error) {
```
bye bye nifty function

kian99 replied:
Sorry about that :cry:

SimoneDutto reviewed (approved):
ty

ale8k commented on internal/jimm/jimm.go:
```diff
@@ -195,9 +196,8 @@ type SSHManager interface {
-	// ControllerInfoFromModelUUID is the method to resolve the address of the controller to contact given the model UUID and
-	// a valid JWT To connect to the controller.
-	ControllerInfoFromModelUUID(ctx context.Context, modelUUID string, user *openfga.User) (ssh.ControllerInfo, error)
+	// DialControllerSSHServer dials the controller hosting the specific model UUID.
```
This is worded weirdly... Uhm maybe something like "dials the hosting controller's SSH server for this model."

ale8k commented on internal/jimm/jimm.go:
```diff
@@ -195,9 +196,8 @@ type SSHManager interface {
-	// a valid JWT To connect to the controller.
-	ControllerInfoFromModelUUID(ctx context.Context, modelUUID string, user *openfga.User) (ssh.ControllerInfo, error)
+	// DialControllerSSHServer dials the controller hosting the specific model UUID.
+	DialControllerSSHServer(ctx context.Context, modelUUID string, user *openfga.User) (*gossh.Client, error)
```
I also really hate how we pass users as extra arguments, we do it so often that we may as well a) put it on context with a specific key and ALWAYS expect a user to be there or b) have a new type which contains the user. Not for this PR but we should discuss this. Passing users as arguments is really ugly and annoying. It also couples our signatures and users to callers.

ale8k commented on internal/jimm/ssh/ssh.go:
```diff
@@ -15,6 +19,9 @@ import (
 )
 
+// jujuSSHDefaultPort is the default port we expect the juju controllers to respond on.
+const jujuSSHDefaultPort = 17022
```
Another discussion for another time, we need a "default values file". Having random scattered vars for defaults doesn't feel good.

ale8k commented on internal/jimm/ssh/ssh.go:
```diff
@@ -112,3 +119,42 @@ func (s *sshManager) ControllerInfoFromModelUUID(ctx context.Context, modelUUID
+// model and dials the controller returning an SSH connection to the controller.
+// We intentionally don't accept user input for the port to disallow the
+// user from probing for open ports on the Juju controller.
+func (s *sshManager) DialControllerSSHServer(ctx context.Context, modelUUID string, user *openfga.User) (*gossh.Client, error) {
```
This is doing too much, we can discover the controller separately and pass the connection info. Would also make testing easier. Otherwise rename it to "DiscoverControllerByModelAndDialControllerSSHServer" if you know what I mean?  Cause really this is a side effect that the function name doesn't portray.

kian99 replied:
I don't agree with this one. It's looking up the controller and dialing. If we want the dialer to be something we define outside and pass in, we can do that in a separate PR. Also the docstring states that it finds the controller by the model.

kian99 replied:
Brought the `controllerInfoFromModelUUID` out of the `DialControllerSSHServer` so that it's not doing two things in one.

ale8k commented on internal/jimm/ssh/ssh.go:
```diff
@@ -112,3 +119,42 @@ func (s *sshManager) ControllerInfoFromModelUUID(ctx context.Context, modelUUID
+	if err != nil {
+		return nil, err
+	}
+	// TODO: Dial the controller and request it's SSH port
```
Make a card please

ale8k commented on internal/jimm/ssh/ssh.go:
```diff
@@ -112,3 +119,42 @@ func (s *sshManager) ControllerInfoFromModelUUID(ctx context.Context, modelUUID
+		client, err = gossh.Dial("tcp", dest, &gossh.ClientConfig{
+			User: "jimm",
+			//nolint:gosec // this will be removed once we handle hostkeys
+			HostKeyCallback: gossh.InsecureIgnoreHostKey(),
```
TODO: don't ignore host key eventually? + Ticket? If we don't put todo in be harder to find and maybe put it on a line separate from the linting line, otherwise might get missed a bit

kian99 replied:
I'll just make a card.

ale8k reviewed (changes requested):
Haven't gone through all but some comments on addressing the side effects should be done first, then will review again

kian99 committed:
chore: Export DialControllerSSHServer again

The DialControllerSSHServer is exported again so that the DialControllerSSHServer is not doing two things in one. Arguably the SSH handler is doing too much and all the logic after the authorisation checks should be moved to a domain server. But that is a larger and separate piece of work.

ale8k commented on internal/ssh/ssh_test.go:
```diff
@@ -245,51 +249,31 @@ func (s *sshSuite) TestSSHJumpDialFail(c *qt.C) {
 	qtsuite.Run(qt.New(t), &sshSuite{})
 }
+
+func inMemoryDial(c *qt.C, listener *bufconn.Listener, config *gossh.ClientConfig) *gossh.Client {
```
lil godoc would go a long way here

ale8k commented on internal/ssh/ssh_test.go:
```diff
@@ -127,18 +120,31 @@ func (s *sshSuite) Init(c *qt.C) {
-			ControllerInfoFromModelUUID_: func(ctx context.Context, modelUUID string, user *openfga.User) (jimmssh.ControllerInfo, error) {
-				if user == userWithAccess {
-					return jimmssh.ControllerInfo{Addresses: []string{""}, JWT: "valid-jwt"}, nil
+			DialControllerSSHServer_: func(ctx context.Context, modelUUID string, user *openfga.User) (*gossh.Client, error) {
```
Snazzy

ale8k commented on internal/testutils/jimmtest/mocks/jimm_ssh_mock.go:
```diff
@@ -5,15 +5,16 @@ package mocks
 )
 
 // SSHManager is an implementation of the SshManager interface.
 type SSHManager struct {
```
Be nice to eventually just gomock interfaces for our services in packages that consume them

ale8k commented on internal/ssh/ssh.go:
```diff
@@ -144,52 +147,51 @@ func directTCPIPHandler(sshManager SSHManager) func(srv *ssh.Server, conn *gossh
-			defer srcDest.Close()
-			defer dstChan.Close()
-			_, err := io.Copy(dstChan, srcDest)
+			defer clientConn.Close()
```
Nice qol

ale8k commented on internal/ssh/ssh.go:
```diff
@@ -131,9 +134,9 @@ func directTCPIPHandler(sshManager SSHManager) func(srv *ssh.Server, conn *gossh
-			d.DestPort = jujuSSHDefaultPort
-		}
+
+		// TODO: Parse destAddr from a virtual hostname.
```
Link ticket here would be good

ale8k commented on internal/ssh/ssh.go:
```diff
@@ -32,9 +32,12 @@ type SSHManager interface {
+	// contact and a valid JWT To authenticate to the controller.
 	ControllerInfoFromModelUUID(ctx context.Context, modelUUID string, user *openfga.User) (jimmssh.ControllerInfo, error)
+
+	// DialControllerSSHServer dials the controller hosting the specific model UUID.
```
Doc is wrong now we're using controller info

ale8k commented on internal/ssh/ssh.go:
```diff
@@ -32,9 +32,12 @@ type SSHManager interface {
 
-	// ControllerInfoFromModelUUID is the method to resolve the address of the controller to contact given the model UUID and
-	// a valid JWT To connect to the controller.
+	// ControllerInfoFromModelUUID uses the given model UUID to return the address of the controller to
```
ControllerInfoFromModelUUID returns controller connection details for the model. It probably shouldn't be returning a JWT but I guess a follow up.

ale8k commented on internal/jimm/ssh/ssh.go:
```diff
@@ -112,3 +119,39 @@ func (s *sshManager) ControllerInfoFromModelUUID(ctx context.Context, modelUUID
+
+// DialControllerSSHServer dials the controller and returns
+// an SSH connection.
+// We don't accept user input for the port to disallow the
```
What does this mean? Surely this isn't the reason we don't allow the specifying of ports?

ale8k reviewed (approved):
lgtm some comments though, please do little godoc ones

kian99 committed:
chore: fix mocks and docstrings

kian99 committed:
Merge branch 'v3' into refactor-ssh-package
